### PR TITLE
Upgrades to RN 59

### PR DESCRIPTION
**Pull request by redreceipt**
## DESCRIPTION

Screen Shot 2019-05-26 at 2 17 40 PM

Screen Shot 2019-05-26 at 2 18 47 PM

### What does this PR do, or why is it needed?

Brings us up to RN 59!

1. Android now has 64-bit support and is incredibly smooth in development.
2. Inline require support (cc @burkeshartsis, could be useful for custom icons)
3. A few components have been deprecated in favor of making RN core very lean. `AsyncStorage` is the only one we need to upgrade eventually, but I tried and the linking is broken ([might need CocoaPods](https://github.com/react-native-community/react-native-async-storage/issues)).
4. ...so when are we gonna try out some hooks???

### What design trade-offs/decisions were made?

Don't think any.

### How do I test this PR?

Probably needs a *thorough* QA session.

### What automated tests did you write?

None.

## TODO:

- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [x] Closes [#606]
- [ ] No new warnings in tests, in storybook, and in-app
- [x] Upload GIF(s) of iOS and Android
- [x] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

**Comment by redreceipt** [comment]
The automatic method didn't work. So far, I've just mimicked the diff-purge here: https://github.com/react-native-community/rn-diff-purge/compare/release/0.58.5..release/0.59.8

I don't _think_ there are any breaking changes to worry about but we'll see. http://facebook.github.io/react-native/blog/2019/03/12/releasing-react-native-059

**Comment by redreceipt** [comment]
error on bundle: 

```
 (node:36030) UnhandledPromiseRejectionWarning: Error: jest-haste-map: Haste module naming collision:
apolloschurchapp:   Duplicate module name: react-native
apolloschurchapp:   Paths: /Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/react-native-linear-gradient/node_modules/react-native/package.json collides with /Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/react-native-iphone-x-helper/node_modules/react-native/package.json
apolloschurchapp: This error is caused by `hasteImpl` returning the same name for different files.
apolloschurchapp:     at setModule (/Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/jest-haste-map/build/index.js:569:17)
apolloschurchapp:     at workerReply (/Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/jest-haste-map/build/index.js:641:9)
apolloschurchapp:     at <anonymous>
apolloschurchapp:     at process._tickCallback (internal/process/next_tick.js:188:7)
apolloschurchapp: (node:36030) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
apolloschurchapp: (node:36030) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by redreceipt** [comment]
Current error is similar: 

```
apolloschurchapp: (node:46725) UnhandledPromiseRejectionWarning: Error: jest-haste-map: Haste module naming collision:
apolloschurchapp:   Duplicate module name: react-native
apolloschurchapp:   Paths: /Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/react-native/package.json collides with /Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/react-native-iphone-x-helper/node_modules/react-native/package.json
apolloschurchapp: This error is caused by `hasteImpl` returning the same name for different files.
apolloschurchapp:     at setModule (/Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/jest-haste-map/build/index.js:569:17)
apolloschurchapp:     at workerReply (/Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/jest-haste-map/build/index.js:641:9)
apolloschurchapp:     at <anonymous>
apolloschurchapp:     at process._tickCallback (internal/process/next_tick.js:188:7)
apolloschurchapp: (node:46725) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
apolloschurchapp: (node:46725) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by redreceipt** [comment]
I'm officially stumped. Current error:

```
apolloschurchapp: error: bundling failed: Error: Unable to resolve module `@apollosproject/ui-storybook` from `/Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/storybook/index.js`: Module `@apollosproject/ui-storybook` does not exist in the Haste module map
apolloschurchapp: This might be related to https://github.com/facebook/react-native/issues/4968
apolloschurchapp: To resolve try the following:
apolloschurchapp:   1. Clear watchman watches: `watchman watch-del-all`.
apolloschurchapp:   2. Delete the `node_modules` folder: `rm -rf node_modules && npm install`.
apolloschurchapp:   3. Reset Metro Bundler cache: `rm -rf /tmp/metro-bundler-cache-*` or `npm start -- --reset-cache`.
apolloschurchapp:   4. Remove haste cache: `rm -rf /tmp/haste-map-react-native-packager-*`.
apolloschurchapp:     at ModuleResolver.resolveDependency (/Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/metro/src/node-haste/DependencyGraph/ModuleResolution.js:183:15)
apolloschurchapp:     at ResolutionRequest.resolveDependency (/Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/metro/src/node-haste/DependencyGraph/ResolutionRequest.js:52:18)
apolloschurchapp:     at DependencyGraph.resolveDependency (/Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/metro/src/node-haste/DependencyGraph.js:283:16)
apolloschurchapp:     at Object.resolve (/Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/metro/src/lib/transformHelpers.js:261:42)
apolloschurchapp:     at dependencies.map.result (/Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/metro/src/DeltaBundler/traverseDependencies.js:399:31)
apolloschurchapp:     at Array.map (<anonymous>)
apolloschurchapp:     at resolveDependencies (/Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/metro/src/DeltaBundler/traverseDependencies.js:396:18)
apolloschurchapp:     at /Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/metro/src/DeltaBundler/traverseDependencies.js:269:33
apolloschurchapp:     at Generator.next (<anonymous>)
apolloschurchapp:     at asyncGeneratorStep (/Users/michael.neeley/Documents/Projects/apollos-prototype/packages/apolloschurchapp/node_modules/metro/src/DeltaBundler/traverseDependencies.js:87:24)
apolloschurchapp:  BUNDLE  [ios, dev] ./index.js ▓░░░░░░░░░░░░░░░ 9.6% (140/451), failed.
```

cc @vinnyjth @conrad-vanl @burkeshartsis

**Comment by vinnyjth** [comment]
@redreceipt I pushed a fix which should get you moving again :)

**Comment by redreceipt** [comment]
`ViewPagerAndroid` has been deprecated and is producing a new warning but I can't find where we are using it.

Screen Shot 2019-05-26 at 1 02 21 PM

File: packages/apolloschurchapp/metro.config.js [hunk 1]
```diff
@@ -12,6 +19,8 @@ const localDeps = [
   path.resolve('..', 'apollos-ui-scripture'),
 ];
 
+console.log(path.resolve('.'), 'LOOK HERE!!!');
```
Comment (vinnyjth) [inline packages/apolloschurchapp/metro.config.js]: I left this in 😂 we need to take it out.

**Comment by burkeshartsis** [comment]
@redreceipt that warning is probably coming from `react-navigation` or something similar `react-native-tab-view` or something like that. I found [this issue](https://github.com/react-navigation/react-navigation/issues/5730) for `react-navigation` but it could just as easily be a different package. Try bumping some of the more obvious packages to more recent releases. New releases might also call out RN59 support for stuff like this. 

Great work btw! SUPER EXCITED for this PR.

**Comment by vinnyjth** [comment]
@redreceipt This PR is .... ABSOLUTELY FANTASTIC.

Thank you much for your thoroughness and attention to detail, especially when documenting. It will be great to have the ability to use hooks 🙌 

I don't know if I will have time to full review this today, but one quick note: Make sure you update the versions of `react` and `react-native` inside each of the `apollos-ui-*`packages. It's not super critical, but will help us avoid module resolve problems down the road.

File: packages/apolloschurchapp/metro.config.js (hunk 1 above)
Comment (redreceipt) [inline packages/apolloschurchapp/metro.config.js]:
```suggestion
```

**Comment by redreceipt** [comment]
@vinnyjth would love to get 5 min of your time to explain your thought process and this commit to get the ball rolling again https://github.com/ApollosProject/apollos-prototype/pull/738/commits/f8b9051545988bfb57f78f68056387c482198f70

**Comment by redreceipt** [comment]
@burkeshartsis I'm more than willing to try and upgrade us to react navigation 3 but do you think that should be a separate PR and maybe we just live with this warning until then? Just reading initial documentation, it doesn't seem complicated but will be a significant change.

**Comment by vinnyjth** [comment]
@redreceipt Future PR for sure. I'll create a github issue.

**Pull request merged**